python/equistore-core/equistore/core/block.py

```diff
@@ -365,10 +365,9 @@ def has_gradient(self, parameter: str) -> bool:
         """
         return parameter in self.gradients_list()
 
-    def gradients(self) -> Generator[Tuple[str, "TensorBlock"], None, None]:
+    def gradients(self) -> Generator[Dict[str, "TensorBlock"], None, None]:
```

Comment:
so then this change is okay right? So we have the same block API on both core and torch?

Reply:
If we want to give up the benefits of a Generator, this should return a `Dict[str, TensorBlock]` directly, and not a generator. But I'm not sure I understand why this is needed. Do you need to get all gradients at the same time?

Reply:
> If we want to give up the benefits of a Generator, this should return a Dict[str, TensorBlock] directly, and not a generator

Yes right.

> If we want to give up the benefits of a Generator, this should return a Dict[str, TensorBlock] directly, and not a generator. But I'm not sure I understand why this is needed. Do you need to get all gradients at the same time?

It is just about the inconsistency between the two APIs. In the core API we use `for key, value in block.gradients()` while in the Torch API does `for key, value in block.gradients().items()`. I can also change the Torch Python API to  return a generator from the Dict it creates that would be probably simpler in terms of refactor.

Reply:
> I can also change the Torch Python API to return a generator from the Dict it creates that would be probably simpler in terms of refactor.

So TorchScript does not support generators, but a list of tuples might be a better choice, allowing to do `for key, value in block.gradients()` in both cases

Reply:
That would require changing this line https://github.com/lab-cosmo/equistore/blob/9053754786f00c9a89151734aa2cbc7c2c91cb7b/equistore-torch/include/equistore/torch/block.hpp#L110

to `std::vector<std::tuple<std::string, TorchTensorBlock>> gradients();` and then propagating the changed everywhere it needs to be. I'm not 100% sure this would work with TorchScript, but I can check quickly

Reply:
> So TorchScript does not support generators, but a list of tuples might be a better choice, allowing to do for key, value in block.gradients() in both cases

Ah okay, then I'll do List of Tuples. Thanks